### ISSUE TITLE
Can't create bot registration

### ISSUE THREAD

**Issue by damaca** [Microsoft/botbuilder-tools#183]
**Describe the bug**
I got an error when I try to create a bot registration:
Unable to provision appid and password for supplied credentials
**To Reproduce**
1. az bot create --resource-group "justtest" --kind "registration" --description "description-of-my-bot" --name "botjustfotest2" -e http://localhost

**Expected behavior**
Create the bot

**Additional context**
I added some prints to the code to see the returned error and I got a 500 error with the following reason:
{"Error":{"Code":"ServiceError","Message":"Cannot create a v2 converged app."}}

**Comment by carlosscastro** [comment]
Hello @damaca, after running the command, you should have got asked to authenticate using a device code. Did you authenticate?

It failed creating your Microsoft app id, which requires successful authentication after issuing the command.

**Comment by damaca** [comment]
yes I did authenticate

**Comment by maxwell-k** [comment]
I have a similar problem when I try:

```
az bot create --resource-group "$resourceGroup" --name "my-bot-name-123485172341234" --kind "webapp" --description "description-of-my-bot"
```

I am already logged in, but get prompted to again. Entering the provided code this time shows "Windows Azure PowerShell" which isn't right - I am using the cross platform CLI. 

If I am experiencing a separate issue, please let me know.

**Comment by aviatoredb** [comment]
Same issue, "Unable to provision appid and password for supplied credentials". Authenticated and all that.
Also downloaded latest CLI as of 5/30

**Comment by swagatmishra2007** [comment]
unfortunately, the code path around the auto provision of msa appid and password is a bit fragile. We are working on making it a bit more robust. 
What is the domain of the resource group that you are using? (eg @gmail.com etc). there are known issues around using @hotmail.com accounts. 

A workaround would be to provision the msaappid manually at apps.dev.microsoft.com/ and pass the created appid and password to the CLI. then it wont prompt for the extra login. i am working on a more permanent fix to this

**Comment by swagatmishra2007** [comment]
@maxwell-k the powershell prompt is a temporary workaround that we are using currently to provision a msa appid and password for the bot. The token obtained when you do an az login doesnt have the correct audience for this, so unfortunately, the second login is necessary . We are working on a long term more seamless experience for this.
As a workaround, you can go to apps.dev.microsoft.com to manually create an app and pass the app credentials to the CLI to avoid this for now.

**Comment by maxwell-k** [comment]
I ended up using a `template.json` / `parameters.json` style deployment rather than the extension. That route also requires manually creating the app registration. 

If you need any testing on the CLI extension route, I'd be happy to help.

**Comment by damaca** [comment]
Could you share the template? 
When I do with the template the application is not created in apps.dev.microsoft.com so the bot doesn't work

**Comment by maxwell-k** [comment]
Template is just the standard one available from "automation options" through when you deploy a chat bot through the portal web UI. It doesn't do anything at apps.dev.microsoft.com. I am not aware of a "scripted" way of registering an app, would like to know about it if there is one.

**Comment by damaca** [comment]
Finally I did it with graph api, you can create that way

**Comment by swagatmishra2007** [comment]
@damaca  can you share the graph api code/script too? since the whole msa appid part seems fragile, i was thinking of removing it from the cli extension and just publishing out samples to script it via graph api. the resultant appid password can then be fed into the cli extension command. 
Please share your script, so that i can put it up in a repo and point folks to it :)

**Comment by damaca** [comment]
Here the functions I'm using to this:

```
oauth_url = 'https://login.microsoftonline.com' 
graph_api_url = 'https://graph.microsoft.com'

def request_token_for_graph(params):
    headers = {

    }

    body = {
        'grant_type': 'client_credentials',
        'client_id': params.client_id,
        'client_secret': params.secret,
        'resource': '{}/'.format(graph_api_url)
    }

    try:
        response = requests.post(oauth_url+'/{}/oauth2/token'.format(params.tenant), data=body)

        if response.status_code != 200:
            print 'There was an error retrieving token: http_code->',response.status,'error->',response.reason
            sys.exit(1)

        return response.json()
    except Exception as e:
        print(e)

def create_msapp(params,token):
    headers = {
        'Authorization': 'Bearer {}'.format(token),
        'Content-Type': 'application/json'
    }

    end_date = datetime.now() + relativedelta(years=20)

    body = {
        'displayName': params.botname,
        'passwordCredentials': [
            {
                'secretText': params.password,
                'endDateTime': end_date.isoformat()
            }
        ]
    }

    try:
        response = requests.post('{}/beta/applications'.format(graph_api_url),data=json.dumps(body),headers=headers)

        if response.status_code != 201:
            print 'There was an error creating app: http_code->',response.status_code,'error->',response.reason
            sys.exit(1)

        return response.json()
    except Exception as e:
        print(e)

```

**Comment by amitbend** [comment]
@damaca What should be the `params.tenant`? applications?
Are you able adding channels etc. with this method?

**Comment by poul-kg** [comment]
Weird, if I specify msa app_id and password via CLI it works. Otherwise I have the same issue:
```
az bot create --resource-group "MyGroup" --name "test-bot-2" --kind "webapp" --description "Bot created via CLI" --lang Node --storage storageGroupName --appid 3bc185c8-f2f4-4bf4-964d-a2aee2281b9c --password 3bc185c8-f2f4-4bf4-964d-a2aee2281b9c
```

here I specified `--appid` & `--password` some random values and it works. But would be great if app registration happens without those values and I can get a JSON response back with app_id and password.

**Comment by poul-kg** [comment]
is there a cli to generate random app_id in proper format?

**Comment by swagatmishra2007** [comment]
@poul-kg the app_id is just a guid. i would suggest using the msa app portal (apps.dev.microsoft.com).
as mentioned above, it's possible to generate this programatically by using the microsoft graph api, but that is in beta, so i havent added it in the cli yet

**Comment by strangecyan** [comment]
@damaca I can create an app against a random tenant using the graph API but that can't be validated against a bot to post to the bot framework. Did you get this to work?

**Comment by QaisAlkhateeb** [comment]
The problem happen when the script try to create new chatbot by the command 'az bot create' without passing appid and password.
for me as a temporary solution I change the command and passing appid and password.
on "node_modules\msbot\bin\msbot-clone-services.js" createBot() function and edit the command to:

`let command = `az bot create -g ${args.groupName} --name ${args.name} --kind webapp --location ${args.location} --insights-location "${args.insightsRegion}" --appid "my_appid" --password "my_password"``

**Comment by ShawnXiao105** [comment]
I still cannot create the bot by adding the parameter appId and appSecret.

**Comment by krajniak** [comment]
You can create suitable msapp with CLI command:
az ad app create --display-name [name] --identifier-uris [uri_in_tenant_domain --password [secret] --available-to-other-tenants true

**Comment by nitinpokharna** [comment]
@krajniak In my Case the bot is not working if I am giving a Random identifier uri for creation of msapp. But when I am deleting the identifier uri in the sdk it is working. Can you please help me how can I pass null identifier uri in az command.

**Comment by krajniak** [comment]
> @krajniak In my Case the bot is not working if I am giving a Random identifier uri for creation of msapp. But when I am deleting the identifier uri in the sdk it is working. Can you please help me how can I pass null identifier uri in az command.

I've tried this, whit different syntax, but was not able to pass a null identifier there. Probably this is a separated bug.

SDK does not work when you pass GUID, hover if you pass a valid uri (it does not have to point to any real address) in tenant domain of you subscription it should work.

I have used something like this: "http://RANDOMLYGENERATEDURI.MYTENANTNAME.onmicrosoft.com/
